### PR TITLE
Audio

### PR DESCRIPTION
**Pull request by sornas**

### CODE REVIEW

File: src/engine/platform/mixer.h [hunk 1]
```diff
@@ -47,6 +48,16 @@ struct Channel {
...
 
+    struct {
+        f32 sum[2];
+        f32 magic;
+        operator bool() const {
+            return magic > 0;
+        }
+    } lowpass = {};
+    void set_lowpass(f32 magic);
+    void remove_lowpass();
```
Comment (FredTheDino) [inline src/engine/platform/mixer.h:59]: This function isn't implemented

File: src/engine/platform/mixer.h [hunk 2]
```diff
@@ -47,6 +48,16 @@ struct Channel {
     void set_delay(f32 feedback, f32 len_seconds);
     void remove_delay();
 
+    struct {
+        f32 sum[2];
+        f32 magic;
```
Comment (FredTheDino) [inline src/engine/platform/mixer.h]: Why is this called magic?

File: src/engine/platform/mixer.h [hunk 3]
```diff
@@ -47,6 +48,16 @@ struct Channel {
...
+        operator bool() const {
+            return magic > 0;
+        }
+    } lowpass = {};
+    void set_lowpass(f32 magic);
```
Comment (FredTheDino) [inline src/engine/platform/mixer.h]: What does the magic passed in here do?

File: src/engine/platform/mixer.h [hunk 4]
```diff
@@ -47,6 +48,16 @@ struct Channel {
     void set_delay(f32 feedback, f32 len_seconds);
     void remove_delay();
```
Comment (FredTheDino) [inline src/engine/platform/mixer.h:49]: This function isn't defined either.

File: src/engine/platform/mixer.cpp [hunk 5]
```diff
@@ -243,6 +258,14 @@ void audio_callback(void* userdata, u8* stream, int len) {
         for (u32 i = 0; i < SAMPLES; i++) {
             output_stream[i] += channel->buffer[(base + i) % CHANNEL_BUFFER_LENGTH];
         }
+
+        for (u32 i = 0; i < SAMPLES; i++) {
+            if (output_stream[i] > SAMPLE_LIMIT) {
+                output_stream[i] = SAMPLE_LIMIT;
+            } else if (output_stream[i] < -SAMPLE_LIMIT) {
+                output_stream[i] = -SAMPLE_LIMIT;
+            }
```
Comment (FredTheDino) [inline src/engine/platform/mixer.cpp]: output_stream[i] = CLAMP(-SAMPLE_LIMIT, SAMPLE_LIMIT, output_stream[i]);

File: src/game/game_main.cpp [hunk 6]
```diff
@@ -90,7 +90,10 @@ void setup() {
         from = *Renderer::get_camera();
     }
     channel = Mixer::fetch_channel(1);
-    channel->set_delay(0.3, 0.2);
+    Mixer::fetch_channel(1)->set_delay(0.3, 0.2);
+    Mixer::fetch_channel(2)->set_delay(0.3, 0.2);
+    Mixer::fetch_channel(1)->set_lowpass(0.5);
+    Mixer::fetch_channel(2)->set_lowpass(0.05);
```
Comment (FredTheDino) [inline src/game/game_main.cpp:96]: This is a nice API! Well done!

**Review by FredTheDino — COMMENTED** [review]
There needs to be a lot more documentation before I would accept this, but the changes look good and it seams to work. Even though the volume seams to be a little bit lower than what it should be when using the low pass filter. A low pass filter doesn't affect the volume does it? But nice code!

**Comment by sornas** [comment]
The volume sounds the same (to me) as when there is no filter active, I've bound W to play the sound without any filters (or delay) to compare it to. Documentation is needed for sure.

File: src/engine/platform/mixer.h [hunk 7]
```diff
@@ -137,6 +137,26 @@ Channel *fetch_channel(u32 channel_id);
...
+// Activates a lowpass filter on the channel with the specified weight. A
+// lowpass filter only lets sounds with a certain frequency or lower through
+// and essentialy filters the high end. This filter is about as basic as it
+// gets and is implemented with a running sum of the previous values. The
+// weight specifies how much the sum is weighted towards the new value,
+// so a lower weight means more filtering. You could probably do some math
+// and figure out / visualize the cutoff for different weights, but knowing
+// that a weight of 1 filters nothing and a weight of 0 filters everything
+// is a good starting point for testing what works for your assets and needs.
+void Channel::set_lowpass(f32 weight);
```
Comment (FredTheDino) [inline src/engine/platform/mixer.h]: This is a bit wordy, If the limits on the filter are 0 and 1 maybe assert the input value is in the given range?

**Review by FredTheDino — APPROVED** [review]
